src/libs/sys-util-rs/src/mount.rs

```diff
@@ -0,0 +1,1073 @@
+    if mountpoint.as_ref().is_empty() || !mountpoint.as_ref().exists() {
+        return Ok(());
+    }
+
+    loop {
+        match umount2(mountpoint.as_ref(), lazy_umount) {
+            Err(e) => {
+                // EINVAL is returned if the target is not a mount point, indicating that we are
+                // done. It can also indicate a few other things (such as invalid flags) which we
+                // unfortunately end up squelching here too.
+                if e.kind() == io::ErrorKind::InvalidInput {
+                    break;
+                } else {
+                    return Err(Error::Umount(mountpoint.as_ref().to_path_buf(), e));
+                }
+            }
+            Ok(()) => (),
+        }
+    }
+
+    Ok(())
+}
+
+// Counterpart of nix::umount2, with support of `UMOUNT_FOLLOW`.
+fn umount2<P: AsRef<Path>>(path: P, lazy_umount: bool) -> std::io::Result<()> {
+    let path_ptr = path.as_ref().as_os_str().as_bytes().as_ptr() as *const i8;
+    let mut flags = UMOUNT_NOFOLLOW;
+    if lazy_umount {
+        flags |= MntFlags::MNT_DETACH.bits();
+    }
+
+    // Safe because parameter is valid and we have checked the reuslt.
+    if unsafe { libc::umount2(path_ptr, flags) } < 0 {
+        Err(io::Error::last_os_error())
+    } else {
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::path::PathBuf;
+
+    #[test]
+    fn test_get_linux_mount_info() {
+        let info = get_linux_mount_info("/sys/fs/cgroup").unwrap();
+
+        assert_eq!(&info.device, "tmpfs");
+        assert_eq!(&info.fs_type, "tmpfs");
+        assert_eq!(&info.path, "/sys/fs/cgroup");
+
+        assert!(matches!(
+            get_linux_mount_info(""),
+            Err(Error::EmptyMountPoint)
+        ));
+        assert!(matches!(
+            get_linux_mount_info("/sys/fs/cgroup/do_not_exist/____hi"),
+            Err(Error::NoMountEntry(_))
+        ));
+    }
+
+    #[test]
+    fn test_create_mount_destination() {
+        let tmpdir = tempfile::tempdir().unwrap();
+        let src = Path::new("/proc/mounts");
+        let mut dst = tmpdir.path().to_owned();
+        dst.push("proc");
+        dst.push("mounts");
+        let dst = create_mount_destination(src, dst.as_path(), tmpdir.path(), "bind").unwrap();
+        let abs_dst = dst.as_ref().canonicalize().unwrap();
+        assert!(abs_dst.is_file());
+
+        let dst = Path::new("/");
+        assert!(matches!(
+            create_mount_destination(src, dst, "/", "bind"),
+            Err(Error::InvalidPath(_))
+        ));
+
+        let src = Path::new("/proc");
+        let dst = Path::new("/proc/mounts");
+        assert!(matches!(
+            create_mount_destination(src, dst, "/", "bind"),
+            Err(Error::InvalidPath(_))
+        ));
+    }
+
+    #[test]
+    fn test_bind_remount_read_only() {
+        // test need root permission
+        if !unistd::getuid().is_root() {
+            println!("test need root permission");
```

Comment:
Hrm, is there really not better way to skip tests in cargo than printing a message and implicitly passing?

Reply:
Thanks for reminder, we can use `#[ignore]` instead of hard coding.